File: ios/Tehilim/Core/Services/ShabbatCalculator.swift
import Foundation

/// Coordonnée géographique (degrés décimaux).
struct GeoCoordinate: Equatable, Codable {
    let latitude: Double
    let longitude: Double
}

/// Une ville pré-enregistrée pour le repli « pas de GPS ».
struct ShabbatCity: Identifiable, Equatable {
    let id: String
    let nameFR: String
    let coordinate: GeoCoordinate
}

/// État Chabbat à un instant donné.
struct ShabbatState: Equatable {
    /// True si l'instant évalué tombe entre l'allumage des bougies (vendredi)
    /// et la sortie de Chabbat / Havdala (samedi soir).
    let isShabbat: Bool
    /// Fin de Chabbat (Havdala) si `isShabbat`, sinon nil.
    let endsAt: Date?
    /// Début du Chabbat (allumage des bougies) — renseigné en Chabbat ET en
    /// pré-Chabbat (pour afficher l'horaire d'entrée à l'avance).
    let startedAt: Date?
    /// Pré-Chabbat : on est dans l'heure qui précède l'entrée. L'écran s'affiche
    /// à l'avance (horaires d'entrée + sortie), sans que Chabbat ait commencé.
    let isPreShabbat: Bool
    /// Prochaine apparition de l'écran (= entrée − 1 h) si on n'est ni en
    /// Chabbat ni en pré-Chabbat — sert à planifier le réveil.
    let nextStartsAt: Date?

    init(isShabbat: Bool, endsAt: Date? = nil, startedAt: Date? = nil,
         isPreShabbat: Bool = false, nextStartsAt: Date? = nil) {
        self.isShabbat = isShabbat
        self.endsAt = endsAt
        self.startedAt = startedAt
        self.isPreShabbat = isPreShabbat
        self.nextStartsAt = nextStartsAt
    }

    /// True si l'écran « Chabbat Chalom » doit s'afficher (pré-Chabbat ou Chabbat).
    var shouldDisplay: Bool { isShabbat || isPreShabbat }
}

/// Calcul du coucher du soleil (algorithme NOAA / Almanac) et de la fenêtre de
/// Chabbat. **Pur** (pas d'UIKit/CoreLocation) → partagé app ↔ widget.
///
/// Règles retenues (cf. décision produit) :
/// - Début : vendredi, coucher du soleil − 18 min (allumage des bougies).
/// - Fin   : samedi, sortie des étoiles (Tzeit) = soleil à 8,5° sous l'horizon
///   (défaut Hebcal « 3 étoiles moyennes »), calculée selon la position.
enum ShabbatCalculator {

    /// Offset allumage des bougies avant le coucher (minutes).
    static let candleLightingOffsetMinutes = 18.0
    /// Havdala = sortie des étoiles : angle de dépression solaire standard
    /// (8,5°, défaut Hebcal). Dépend de la position et de la date.
    static let havdalahDepressionDegrees = 8.5
    /// Repli quand le soleil n'atteint pas 8,5° sous l'horizon (hautes
    /// latitudes en été) : coucher + 72 min.
    static let havdalahFallbackOffsetMinutes = 72.0
    /// L'écran s'affiche dès 1 h avant l'entrée (pré-Chabbat) pour informer
    /// l'utilisateur des horaires d'entrée et de sortie.
    static let preShabbatLeadSeconds: TimeInterval = 3600

    /// Liste de villes pour le repli sans GPS (id stable, ne pas renommer).
    static let cities: [ShabbatCity] = [
        .init(id: "jerusalem", nameFR: "Jérusalem", coordinate: .init(latitude: 31.7683, longitude: 35.2137)),
        .init(id: "telaviv",   nameFR: "Tel Aviv",  coordinate: .init(latitude: 32.0853, longitude: 34.7818)),
        .init(id: "paris",     nameFR: "Paris",      coordinate: .init(latitude: 48.8566, longitude: 2.3522)),
        .init(id: "marseille", nameFR: "Marseille",  coordinate: .init(latitude: 43.2965, longitude: 5.3698)),
        .init(id: "lyon",      nameFR: "Lyon",        coordinate: .init(latitude: 45.7640, longitude: 4.8357)),
        .init(id: "nice",      nameFR: "Nice",        coordinate: .init(latitude: 43.7102, longitude: 7.2620)),
        .init(id: "strasbourg", nameFR: "Strasbourg", coordinate: .init(latitude: 48.5734, longitude: 7.7521)),
        .init(id: "toulouse",  nameFR: "Toulouse",    coordinate: .init(latitude: 43.6047, longitude: 1.4442)),
        .init(id: "london",    nameFR: "Londres",     coordinate: .init(latitude: 51.5074, longitude: -0.1278)),
        .init(id: "brussels",  nameFR: "Bruxelles",   coordinate: .init(latitude: 50.8503, longitude: 4.3517)),
        .init(id: "geneva",    nameFR: "Genève",      coordinate: .init(latitude: 46.2044, longitude: 6.1432)),
        .init(id: "montreal",  nameFR: "Montréal",    coordinate: .init(latitude: 45.5019, longitude: -73.5674)),
        .init(id: "newyork",   nameFR: "New York",    coordinate: .init(latitude: 40.7128, longitude: -74.0060)),
        .init(id: "losangeles", nameFR: "Los Angeles", coordinate: .init(latitude: 34.0522, longitude: -118.2437))
    ]

    static func city(id: String) -> ShabbatCity? { cities.first { $0.id == id } }

    // MARK: - État Chabbat

    /// Évalue l'état Chabbat à `now` pour la position `coord`.
    static func state(now: Date, coordinate: GeoCoordinate,
                      timeZone: TimeZone = .current) -> ShabbatState {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = timeZone

        let weekday = cal.component(.weekday, from: now) // 1=dim … 6=ven, 7=sam

        // Fenêtre d'affichage = [entrée − 1 h, havdala]. Bloquant à partir de
        // l'entrée (bougies) ; informatif (pré-Chabbat) dans l'heure avant.
        if weekday == 6 { // vendredi
            if let friday = dateOnly(now, cal: cal),
               let saturday = cal.date(byAdding: .day, value: 1, to: friday),
               let candle = candleLighting(on: friday, coordinate: coordinate, cal: cal),
               let havdalah = havdalah(on: saturday, coordinate: coordinate, cal: cal) {
                if now >= candle {
                    return ShabbatState(isShabbat: true, endsAt: havdalah, startedAt: candle)
                }
                let preStart = candle.addingTimeInterval(-preShabbatLeadSeconds)
                if now >= preStart {
                    return ShabbatState(isShabbat: false, endsAt: havdalah,
                                        startedAt: candle, isPreShabbat: true)
                }
                return ShabbatState(isShabbat: false, nextStartsAt: preStart)
            }
        } else if weekday == 7 { // samedi
            if let saturday = dateOnly(now, cal: cal),
               let friday = cal.date(byAdding: .day, value: -1, to: saturday),
               let havdalah = havdalah(on: saturday, coordinate: coordinate, cal: cal) {
                if now <= havdalah {
                    // startedAt = allumage des bougies de la veille (vendredi).
                    let candle = candleLighting(on: friday, coordinate: coordinate, cal: cal)
                    return ShabbatState(isShabbat: true, endsAt: havdalah, startedAt: candle)
                }
            }
        }

        // Hors fenêtre : prochaine apparition de l'écran = entrée − 1 h.
        let next = nextCandleLighting(after: now, coordinate: coordinate, cal: cal)
        return ShabbatState(isShabbat: false,
                            nextStartsAt: next?.addingTimeInterval(-preShabbatLeadSeconds))
    }

    private static func nextCandleLighting(after now: Date, coordinate: GeoCoordinate,
                                           cal: Calendar) -> Date? {
        guard let today = dateOnly(now, cal: cal) else { return nil }
        // Cherche le prochain vendredi (0…7 jours) dont l'allumage est > now.
        for offset in 0...7 {
            guard let day = cal.date(byAdding: .day, value: offset, to: today) else { continue }
            if cal.component(.weekday, from: day) == 6,
               let candle = candleLighting(on: day, coordinate: coordinate, cal: cal),
               candle > now {
                return candle
            }
        }
        return nil
    }

    private static func candleLighting(on friday: Date, coordinate: GeoCoordinate,
                                       cal: Calendar) -> Date? {
        guard let s = sunset(on: friday, coordinate: coordinate, cal: cal) else { return nil }
        return s.addingTimeInterval(-candleLightingOffsetMinutes * 60)
    }

    private static func havdalah(on saturday: Date, coordinate: GeoCoordinate,
                                 cal: Calendar) -> Date? {
        // Sortie des étoiles : soleil à 8,5° sous l'horizon (zénith 98,5°).
        if let tzeit = eveningEvent(on: saturday, coordinate: coordinate, cal: cal,
                                    zenith: 90.0 + havdalahDepressionDegrees) {
            return tzeit
        }
        // Repli (le soleil ne descend pas jusqu'à 8,5° près du solstice à
        // haute latitude) : coucher + 72 min.
        guard let s = sunset(on: saturday, coordinate: coordinate, cal: cal) else { return nil }
        return s.addingTimeInterval(havdalahFallbackOffsetMinutes * 60)
    }

    private static func dateOnly(_ date: Date, cal: Calendar) -> Date? {
        cal.startOfDay(for: date)
    }

    // MARK: - Coucher du soleil (NOAA / Almanac)

    /// Coucher du soleil pour la date civile de `day` (dans `cal.timeZone`) à
    /// la position donnée. Retourne l'instant absolu (Date), ou nil aux
    /// latitudes polaires où le soleil ne se couche pas ce jour-là.
    static func sunset(on day: Date, coordinate: GeoCoordinate, cal: Calendar) -> Date? {
        // Coucher du soleil = zénith officiel 90°50' (réfraction + rayon).
        eveningEvent(on: day, coordinate: coordinate, cal: cal, zenith: 90.833)
    }

    /// Événement solaire du soir pour un `zenith` donné (90,833 = coucher ;
    /// > 90 = crépuscule / sortie des étoiles selon l'angle de dépression).
    static func eveningEvent(on day: Date, coordinate: GeoCoordinate, cal: Calendar,
                             zenith: Double) -> Date? {
        let comps = cal.dateComponents([.year, .month, .day], from: day)
        guard let year = comps.year, let month = comps.month, let dayNum = comps.day else { return nil }
        return sunsetUTC(year: year, month: month, day: dayNum,
                         latitude: coordinate.latitude, longitude: coordinate.longitude,
                         timeZone: cal.timeZone, zenith: zenith)
    }

    private static func deg2rad(_ d: Double) -> Double { d * .pi / 180 }
    private static func rad2deg(_ r: Double) -> Double { r * 180 / .pi }
    private static func mod(_ a: Double, _ b: Double) -> Double { let r = a.truncatingRemainder(dividingBy: b); return r < 0 ? r + b : r }

    /// Algorithme « Sunrise/Sunset » de l'Almanac (NOAA), précision ~1 min.
    private static func sunsetUTC(year: Int, month: Int, day: Int,
                                  latitude: Double, longitude: Double,
                                  timeZone: TimeZone, zenith: Double = 90.833) -> Date? {
        // 1. Jour de l'année
        let N1 = floor(275.0 * Double(month) / 9.0)
        let N2 = floor((Double(month) + 9.0) / 12.0)
        let N3 = 1.0 + floor((Double(year) - 4.0 * floor(Double(year) / 4.0) + 2.0) / 3.0)
        let N = N1 - (N2 * N3) + Double(day) - 30.0

        let lngHour = longitude / 15.0
        // Approximation de l'heure (sunset → 18h)
        let t = N + ((18.0 - lngHour) / 24.0)

        // Anomalie moyenne du soleil
        let M = (0.9856 * t) - 3.289
        // Longitude vraie
        var L = M + (1.916 * sin(deg2rad(M))) + (0.020 * sin(deg2rad(2 * M))) + 282.634
        L = mod(L, 360.0)
        // Ascension droite
        var RA = rad2deg(atan(0.91764 * tan(deg2rad(L))))
        RA = mod(RA, 360.0)
        // Même quadrant que L
        let Lquadrant = floor(L / 90.0) * 90.0
        let RAquadrant = floor(RA / 90.0) * 90.0
        RA = (RA + (Lquadrant - RAquadrant)) / 15.0
        // Déclinaison
        let sinDec = 0.39782 * sin(deg2rad(L))
        let cosDec = cos(asin(sinDec))
        // Angle horaire — `zenith` passé en paramètre (90,833 = coucher,
        // 98,5 = sortie des étoiles à 8,5°).
        let cosH = (cos(deg2rad(zenith)) - (sinDec * sin(deg2rad(latitude)))) /
                   (cosDec * cos(deg2rad(latitude)))
        if cosH > 1 || cosH < -1 { return nil } // pas de coucher (polaire)
        // Sunset → H = acos(cosH)
        let H = rad2deg(acos(cosH)) / 15.0
        let T = H + RA - (0.06571 * t) - 6.622
        let UT = mod(T - lngHour, 24.0)

        // Construit l'instant : UTC du jour (year,month,day) à UT heures.
        let hour = Int(floor(UT))
        let minuteFull = (UT - Double(hour)) * 60.0
        let minute = Int(floor(minuteFull))
        let second = Int((minuteFull - Double(minute)) * 60.0)

        var utc = Calendar(identifier: .gregorian)
        utc.timeZone = TimeZone(identifier: "UTC")!
        var dc = DateComponents()
        dc.year = year; dc.month = month; dc.day = day
        dc.hour = hour; dc.minute = minute; dc.second = second
        guard let candidate = utc.date(from: dc) else { return nil }

        // Correction de bord : aux longitudes où le coucher tombe de l'autre
        // côté de minuit UTC, l'instant calculé peut être décalé d'un jour.
        // On choisit le décalage (−1/0/+1 j) dont la date LOCALE colle au jour
        // visé.
        var targetCal = Calendar(identifier: .gregorian)
        targetCal.timeZone = timeZone
        for shift in [0, 1, -1] {
            let c = candidate.addingTimeInterval(Double(shift) * 86400.0)
            let lc = targetCal.dateComponents([.year, .month, .day], from: c)
            if lc.year == year && lc.month == month && lc.day == day {
                return c
            }
        }
        return candidate
    }
}

File: android/app/src/main/java/com/david/tehilim/core/service/ShabbatCalculator.kt
package com.david.tehilim.core.service

import java.util.Calendar
import java.util.Date
import java.util.GregorianCalendar
import java.util.TimeZone
import kotlin.math.acos
import kotlin.math.asin
import kotlin.math.atan
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.sin
import kotlin.math.tan

/** Coordonnée géographique (degrés décimaux). */
data class GeoCoordinate(val latitude: Double, val longitude: Double)

/** Ville pré-enregistrée pour le repli « pas de GPS ». */
data class ShabbatCity(val id: String, val nameFR: String, val coordinate: GeoCoordinate)

/** État Chabbat à un instant donné (mirror iOS). */
data class ShabbatState(
    val isShabbat: Boolean,
    val endsAt: Date?,
    val nextStartsAt: Date?,
    /** Début du Chabbat (allumage des bougies) — renseigné en Chabbat ET en
     *  pré-Chabbat (pour afficher l'horaire d'entrée à l'avance). */
    val startedAt: Date? = null,
    /** Pré-Chabbat : dans l'heure qui précède l'entrée (écran informatif). */
    val isPreShabbat: Boolean = false
) {
    /** True si l'écran « Chabbat Chalom » doit s'afficher. */
    val shouldDisplay: Boolean get() = isShabbat || isPreShabbat
}

/**
 * Mirror Android du ShabbatCalculator.swift.
 *
 * Calcul du coucher du soleil (algorithme NOAA / Almanac) et de la fenêtre de
 * Chabbat. Pur (aucune dépendance Android) → partageable app ↔ widget Glance.
 *
 * - Début : vendredi, coucher du soleil − 18 min (allumage des bougies).
 * - Fin   : samedi, sortie des étoiles (Tzeit) = soleil à 8,5° sous l'horizon
 *   (défaut Hebcal « 3 étoiles moyennes »), calculée selon la position.
 */
object ShabbatCalculator {

    const val CANDLE_LIGHTING_OFFSET_MIN = 18.0
    /** Havdala = sortie des étoiles : angle de dépression solaire standard
     *  (8,5°, défaut Hebcal). Dépend de la position et de la date. */
    const val HAVDALAH_DEPRESSION_DEG = 8.5
    /** Repli quand le soleil n'atteint pas 8,5° (hautes latitudes en été). */
    const val HAVDALAH_FALLBACK_OFFSET_MIN = 72.0
    /** L'écran s'affiche dès 1 h avant l'entrée (pré-Chabbat). */
    const val PRE_SHABBAT_LEAD_MS = 3_600_000L

    val cities: List<ShabbatCity> = listOf(
        ShabbatCity("jerusalem", "Jérusalem", GeoCoordinate(31.7683, 35.2137)),
        ShabbatCity("telaviv", "Tel Aviv", GeoCoordinate(32.0853, 34.7818)),
        ShabbatCity("paris", "Paris", GeoCoordinate(48.8566, 2.3522)),
        ShabbatCity("marseille", "Marseille", GeoCoordinate(43.2965, 5.3698)),
        ShabbatCity("lyon", "Lyon", GeoCoordinate(45.7640, 4.8357)),
        ShabbatCity("nice", "Nice", GeoCoordinate(43.7102, 7.2620)),
        ShabbatCity("strasbourg", "Strasbourg", GeoCoordinate(48.5734, 7.7521)),
        ShabbatCity("toulouse", "Toulouse", GeoCoordinate(43.6047, 1.4442)),
        ShabbatCity("london", "Londres", GeoCoordinate(51.5074, -0.1278)),
        ShabbatCity("brussels", "Bruxelles", GeoCoordinate(50.8503, 4.3517)),
        ShabbatCity("geneva", "Genève", GeoCoordinate(46.2044, 6.1432)),
        ShabbatCity("montreal", "Montréal", GeoCoordinate(45.5019, -73.5674)),
        ShabbatCity("newyork", "New York", GeoCoordinate(40.7128, -74.0060)),
        ShabbatCity("losangeles", "Los Angeles", GeoCoordinate(34.0522, -118.2437))
    )

    fun city(id: String): ShabbatCity? = cities.firstOrNull { it.id == id }

    // MARK: - État Chabbat

    fun state(
        now: Date,
        coordinate: GeoCoordinate,
        timeZone: TimeZone = TimeZone.getDefault()
    ): ShabbatState {
        val cal = GregorianCalendar(timeZone)
        cal.time = now
        val weekday = cal.get(Calendar.DAY_OF_WEEK)

        if (weekday == Calendar.FRIDAY) {
            val friday = startOfDay(now, timeZone)
            val saturday = addDays(friday, 1, timeZone)
            val candle = candleLighting(friday, coordinate, timeZone)
            val havdalah = havdalah(saturday, coordinate, timeZone)
            if (candle != null && havdalah != null) {
                if (!now.before(candle)) return ShabbatState(true, havdalah, null, startedAt = candle)
                val preStart = Date(candle.time - PRE_SHABBAT_LEAD_MS)
                if (!now.before(preStart)) {
                    return ShabbatState(false, havdalah, null, startedAt = candle, isPreShabbat = true)
                }
                return ShabbatState(false, null, preStart)
            }
        } else if (weekday == Calendar.SATURDAY) {
            val saturday = startOfDay(now, timeZone)
            val friday = addDays(saturday, -1, timeZone)
            val candle = candleLighting(friday, coordinate, timeZone)
            val havdalah = havdalah(saturday, coordinate, timeZone)
            if (candle != null && havdalah != null && !now.after(havdalah)) {
                return ShabbatState(true, havdalah, null, startedAt = candle)
            }
        }
        val next = nextCandleLighting(now, coordinate, timeZone)
        return ShabbatState(false, null, next?.let { Date(it.time - PRE_SHABBAT_LEAD_MS) })
    }

    private fun nextCandleLighting(now: Date, coordinate: GeoCoordinate, tz: TimeZone): Date? {
        val today = startOfDay(now, tz)
        for (offset in 0..7) {
            val day = addDays(today, offset, tz)
            val cal = GregorianCalendar(tz).apply { time = day }
            if (cal.get(Calendar.DAY_OF_WEEK) == Calendar.FRIDAY) {
                val candle = candleLighting(day, coordinate, tz)
                if (candle != null && candle.after(now)) return candle
            }
        }
        return null
    }

    private fun candleLighting(friday: Date, coordinate: GeoCoordinate, tz: TimeZone): Date? {
        val s = sunset(friday, coordinate, tz) ?: return null
        return Date(s.time - (CANDLE_LIGHTING_OFFSET_MIN * 60_000).toLong())
    }

    private fun havdalah(saturday: Date, coordinate: GeoCoordinate, tz: TimeZone): Date? {
        // Sortie des étoiles : soleil à 8,5° sous l'horizon (zénith 98,5°).
        eveningEvent(saturday, coordinate, tz, 90.0 + HAVDALAH_DEPRESSION_DEG)?.let { return it }
        // Repli (le soleil ne descend pas à 8,5° près du solstice à haute
        // latitude) : coucher + 72 min.
        val s = sunset(saturday, coordinate, tz) ?: return null
        return Date(s.time + (HAVDALAH_FALLBACK_OFFSET_MIN * 60_000).toLong())
    }

    private fun startOfDay(date: Date, tz: TimeZone): Date {
        val cal = GregorianCalendar(tz).apply {
            time = date
            set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
        }
        return cal.time
    }

    private fun addDays(date: Date, days: Int, tz: TimeZone): Date {
        val cal = GregorianCalendar(tz).apply { time = date; add(Calendar.DAY_OF_MONTH, days) }
        return cal.time
    }

    // MARK: - Coucher du soleil (NOAA / Almanac)

    fun sunset(day: Date, coordinate: GeoCoordinate, tz: TimeZone): Date? =
        eveningEvent(day, coordinate, tz, 90.833)

    /**
     * Événement solaire du soir pour un `zenith` donné (90,833 = coucher ;
     * > 90 = crépuscule / sortie des étoiles selon l'angle de dépression).
     */
    fun eveningEvent(day: Date, coordinate: GeoCoordinate, tz: TimeZone, zenith: Double): Date? {
        val cal = GregorianCalendar(tz).apply { time = day }
        return sunsetUTC(
            cal.get(Calendar.YEAR),
            cal.get(Calendar.MONTH) + 1, // Calendar.MONTH is 0-based
            cal.get(Calendar.DAY_OF_MONTH),
            coordinate.latitude,
            coordinate.longitude,
            tz,
            zenith
        )
    }

    private fun deg2rad(d: Double) = d * Math.PI / 180.0
    private fun rad2deg(r: Double) = r * 180.0 / Math.PI
    private fun mod(a: Double, b: Double): Double { val r = a % b; return if (r < 0) r + b else r }

    private fun sunsetUTC(
        year: Int, month: Int, day: Int,
        latitude: Double, longitude: Double, timeZone: TimeZone, zenith: Double = 90.833
    ): Date? {
        val n1 = floor(275.0 * month / 9.0)
        val n2 = floor((month + 9.0) / 12.0)
        val n3 = 1.0 + floor((year - 4.0 * floor(year / 4.0) + 2.0) / 3.0)
        val n = n1 - (n2 * n3) + day - 30.0

        val lngHour = longitude / 15.0
        val t = n + ((18.0 - lngHour) / 24.0)
        val m = (0.9856 * t) - 3.289
        var l = m + (1.916 * sin(deg2rad(m))) + (0.020 * sin(deg2rad(2 * m))) + 282.634
        l = mod(l, 360.0)
        var ra = rad2deg(atan(0.91764 * tan(deg2rad(l))))
        ra = mod(ra, 360.0)
        val lQuadrant = floor(l / 90.0) * 90.0
        val raQuadrant = floor(ra / 90.0) * 90.0
        ra = (ra + (lQuadrant - raQuadrant)) / 15.0
        val sinDec = 0.39782 * sin(deg2rad(l))
        val cosDec = cos(asin(sinDec))
        // `zenith` passé en paramètre (90,833 = coucher, 98,5 = tzeit à 8,5°).
        val cosH = (cos(deg2rad(zenith)) - (sinDec * sin(deg2rad(latitude)))) /
            (cosDec * cos(deg2rad(latitude)))
        if (cosH > 1 || cosH < -1) return null
        val h = rad2deg(acos(cosH)) / 15.0
        val localT = h + ra - (0.06571 * t) - 6.622
        val ut = mod(localT - lngHour, 24.0)

        val hour = floor(ut).toInt()
        val minuteFull = (ut - hour) * 60.0
        val minute = floor(minuteFull).toInt()
        val second = ((minuteFull - minute) * 60.0).toInt()

        val utc = GregorianCalendar(TimeZone.getTimeZone("UTC")).apply {
            clear()
            set(year, month - 1, day, hour, minute, second)
        }
        val candidate = utc.time

        // Correction de bord : choisit le décalage (−1/0/+1 j) dont la date
        // LOCALE colle au jour visé (cas des longitudes traversant minuit UTC).
        val target = GregorianCalendar(timeZone)
        for (shift in intArrayOf(0, 1, -1)) {
            val c = Date(candidate.time + shift.toLong() * 86_400_000L)
            target.time = c
            if (target.get(Calendar.YEAR) == year &&
                target.get(Calendar.MONTH) + 1 == month &&
                target.get(Calendar.DAY_OF_MONTH) == day
            ) return c
        }
        return candidate
    }
}
